### PR TITLE
Make default parameterized display name template configurable

### PR DESCRIPTION
**Pull request by chia7712**
Introduce `junit.jupiter.params.displayname.default` configuration parameter to configure the default display name template for all parameterized tests.

## Overview

according to the comment (https://github.com/junit-team/junit5/pull/2526#issuecomment-760898928)

The new configuration ```junit.jupiter.params.displayname.default``` enables us to configure the display name for all parameterized test. With this new configuration, we don't need to write duplicate pattern (like ```{displayName} - {arguments}```) to all test case.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [ ] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestExtension.java
Comment (marcphilipp) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestExtension.java:42]: I know I said "template" in the other issue. Looking at the sources of `@ParameterizedTest` now I think "pattern" would be better. Looking at other configuration paramters, though, I think `junit.jupiter.params.displayname.default` would be more consistent.

@junit-team/junit-lambda WDYT?

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestExtension.java
Comment (marcphilipp) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestExtension.java:136]: Please only call `trim()` once and add validation (with an appropriate message) for the config parameter along with a test case each for both scenarios.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:1600]: Actually, the logic should be inverted: `ParameterizedTest.name()` should override the config parameter. Please also document the config parameter in the Javadoc of `ParameterizedTest.name()`.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (sbrannen) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:1600]: That's correct: `ParameterizedTest.name()` must override the config parameter in all cases.

But.... to make that possible, we will need to change the `default` value of `ParameterizedTest.name()`.

If we leave it declared as `String name() default DEFAULT_DISPLAY_NAME;` (i.e., the status quo), it will be impossible to know if the user explicitly set the `name` attribute to the value of `DEFAULT_DISPLAY_NAME` intentionally.

Thus, we will need to change the default of the `name` attribute from `DEFAULT_DISPLAY_NAME` to something like `"{default_display_name}"`.

If `"{default_display_name}"` is returned when invoking `name()`, we will:

1. Look up the new config param, and if it has been set, use it.
2. If the new config param has not been set, we will use the value of the `DEFAULT_DISPLAY_NAME` constant for backwards compatibility.

Otherwise, the user explicitly supplied a pattern for the `name` attribute (or their code was compiled against a previous version of the API with `DEFAULT_DISPLAY_NAME` set as the default value for `name`), and we use whatever explicit pattern is set.

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestExtension.java
Comment (sbrannen) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestExtension.java:42]: The term "pattern" is consistent with the existing documentation for `DEFAULT_DISPLAY_NAME`:

> Default display name pattern for the current invocation of a `@ParameterizedTest` method: `"[{index}] {argumentsWithNames}"`

Comment (sbrannen) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestExtension.java:42]: > Looking at other configuration paramters, though, I think `junit.jupiter.params.displayname.default` would be more consistent.

Yes, let's go with `junit.jupiter.params.displayname.default`.

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestExtension.java
Comment (sbrannen) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestExtension.java:136]: As mentioned in https://github.com/junit-team/junit5/pull/2532/files#r559023221 and subsequent comments, the priority needs to be reversed for backwards compatibility and local overrides of the new global default.

**Review by sbrannen — CHANGES_REQUESTED** [review]

**Comment by chia7712** [comment]
@sbrannen @marcphilipp Thanks for all your suggestions. I have addressed your comments. Please take a look.

**Review by sbrannen — APPROVED** [review]
Thanks for making the requested changes.

I think this PR looks pretty good as is.

The team can take over any final polishing before merging into `main`.

But let's first see what the rest of the team has to say.

**Comment by chia7712** [comment]
@sbrannen @marcphilipp any feedback? I'd like to address your comment ASAP since it would be better if this patch can be included by junit 5.8 :)

**Comment by sbrannen** [comment]
> @sbrannen @marcphilipp any feedback? 

I approved the PR.

There are some minor improvements to be made to the documentation, etc., but those can be addressed by a core committer once the PR has been merged into `main`.

So we are waiting on approval from @marcphilipp.

> I'd like to address your comment ASAP since it would be better if this patch can be included by junit 5.8 :)

It looks like this can potentially make it into 5.8 M1.

**Comment by chia7712** [comment]
fix conflicting file

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:1664]:
```suggestion
3. `DEFAULT_DISPLAY_NAME` constant defined in `@ParameterizedTest`
```

**Review by marcphilipp — APPROVED** [review]

**Pull request merged**
